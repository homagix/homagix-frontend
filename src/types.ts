export type AppError = {
  message: string
  details?: unknown
}

export type Item = {
  id: string
  amount: number
}

export type Dish = {
  id: string
  name: string
  image: string
  recipe?: string
  source?: string
  last?: string
  items: Item[]
  alwaysOnList: boolean
}

export type Ingredient = {
  id: string
  amount: number
  unit: string
  name: string
}
