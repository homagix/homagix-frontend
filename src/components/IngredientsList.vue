<script setup lang="ts">
import type { Ingredient } from "@/types"

defineProps<{ ingredients: Ingredient[] }>()
</script>

<template>
  <div class="ingredient-list">
    <template v-for="ingredient in ingredients" :key="ingredient.id">
      <span>{{ ingredient.amount }}</span>
      <span>{{ ingredient.unit }}</span>
      <span>{{ ingredient.name }}</span>
    </template>
  </div>
</template>

<style lang="scss" scoped>
.ingredient-list {
  display: inline-grid;
  grid-template-rows: auto;
  grid-template-columns: 1fr 1fr 6fr;
  border: 1px solid #888;
  padding: 3px;
  align-content: flex-start;
  min-height: 300px;
  min-width: 200px;

  @media (max-width: 599px) {
    width: 100%;
  }

  > * {
    padding: 3px;
  }

  > *:nth-child(3n-2) {
    text-align: right;
  }
}
</style>
