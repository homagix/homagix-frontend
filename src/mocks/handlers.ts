import { rest } from "msw"

export const handlers = [
  rest.get("http://localhost:8200/api/dishes", (req, res, ctx) => {
    return res(ctx.json({ dishes: [{ id: 1, name: "Pizza" }]}))
  }),
]
