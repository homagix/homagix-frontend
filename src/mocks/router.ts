import { defineComponent } from "vue"
import { createWebHistory } from "vue-router"
import { createRouter } from "vue-router"

export function createRouterMock(component: ReturnType<typeof defineComponent>, extraPaths: string[] = []) {
  const history = createWebHistory()
  const routes = ["/", ...extraPaths].map(path => ({ path, component }))
  return createRouter({ history, routes })
}
