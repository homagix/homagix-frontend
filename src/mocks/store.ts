import { vi } from "vitest"
import { createStore } from "vuex"

export function createStoreMock(state: object = {}, options: { actions?: string[]; mutations?: string[] } = {}) {
  function createMocksRecord(names?: string[]) {
    if (!names) {
      return undefined
    }
    return Object.assign({}, ...names.map(name => ({ [name]: vi.fn() })))
  }

  const actionMocks = createMocksRecord(options.actions)
  const mutationMocks = createMocksRecord(options.mutations)
  const store = createStore({
    state() {
      return state
    },
    actions: actionMocks,
    mutations: mutationMocks,
  })
  return { store, actionMocks, mutationMocks }
}
