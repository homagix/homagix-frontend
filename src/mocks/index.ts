import { mount, shallowMount } from "@vue/test-utils"
import { createRouterMock } from "./router"
import { createStoreMock, StoreMockOptions } from "./store"
import Oruga from "@oruga-ui/oruga-next"

type MountOptions = {
  shallow?: boolean
  props?: Record<string, any>
  state?: object
  storeOptions?: StoreMockOptions
  extraPaths?: string[]
}

export function mountComponent<C>(component: C, options: MountOptions) {
  const { store, actionMocks, mutationMocks } = createStoreMock(options.state || {}, options.storeOptions || {})
  const router = createRouterMock(component, options.extraPaths || [])
  const m = options.shallow === undefined || options.shallow ? shallowMount : mount
  const wrapper = m(component, { props: options.props || {}, global: { plugins: [store, router, Oruga] } })

  return { wrapper, store, router, actionMocks, mutationMocks }
}
