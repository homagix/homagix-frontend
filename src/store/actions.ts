import { Context, MutationType, ActionType } from "@/store"
import { AppError } from "@/types"
import { loadDishes } from "@/api/dishes"
import { loadIngredients } from "@/api/ingredients"
import { logoutUser, registerUser } from "@/api/user"

type DropFirst<T extends unknown[]> = T extends [unknown, ...infer U] ? U : never

function catchErrors<T extends (context: Context, ...args: never[]) => ReturnType<T>>(
  func: T,
  message: string
): (context: Context, ...funcArgs: DropFirst<Parameters<T>>) => Promise<void> {
  return async (context, ...args) => {
    try {
      await func(context, ...args)
    } catch (details) {
      context.dispatch(ActionType.ERROR_OCCURED, { message, details })
    }
  }
}

export default {
  ERROR_OCCURED(context: Context, error: AppError) {
    context.commit(MutationType.SET_ERROR, error)
  },

  CLEAR_ERROR(context: Context) {
    context.commit(MutationType.SET_ERROR, null)
  },

  LOAD_DISHES: catchErrors(
    async context => context.commit(MutationType.LOADED_DISHES, await loadDishes()),
    "Rezepte konnten nicht geladen werden"
  ),

  LOAD_INGREDIENTS: catchErrors(
    async context => context.commit(MutationType.LOADED_INGREDIENTS, await loadIngredients()),
    "Zutaten konnten nicht geladen werden"
  ),

  REGISTER_USER: catchErrors(
    async (context, name: string) => context.commit(MutationType.SET_USER, await registerUser(name as string)),
    "Die Registrierung war leider nicht erfolgreich. Bitte noch einmal probieren!"
  ),

  LOGOUT: catchErrors(async context => {
    await logoutUser()
    context.commit(MutationType.SET_USER, null)
  }, "Logout fehlgeschlagen"),
}
