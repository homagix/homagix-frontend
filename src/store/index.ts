import { Store as VuexStore, useStore as baseUseStore, createStore, ActionContext } from "vuex"
import mutations from "@/store/mutations"
import actions from "@/store/actions"
import getters from "@/store/getters"
import { AppError, Dish, Ingredient } from "@/types"
import { state as auth } from "@/auth"

const state = {
  error: null as AppError | null,
  dishes: [] as Dish[],
  ingredients: [] as Ingredient[],
  ...auth,
}

const store = createStore({
  strict: import.meta.env.NODE_ENV !== "production",
  state,
  mutations,
  actions,
  getters,
})

function listKeys(obj: object) {
  return Object.assign({}, ...Object.keys(obj).map(key => ({ [key]: key })))
}

export default store
export type State = typeof state
export type Mutations = typeof mutations
export type Actions = typeof actions
export type Getters = typeof getters
export const MutationType = listKeys(mutations) as Record<keyof Mutations, keyof Mutations>
export const ActionType = listKeys(actions) as Record<keyof Actions, keyof Actions>

type MutationsProp = {
  commit<K extends keyof Mutations>(key: K, payload: Parameters<Mutations[K]>[1]): ReturnType<Mutations[K]>
}

export type Context = Omit<ActionContext<State, State>, "commit"> & MutationsProp
export type Store = VuexStore<State>

export function useStore(): Store {
  return baseUseStore()
}
