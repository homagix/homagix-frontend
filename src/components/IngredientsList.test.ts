import { Ingredient } from "./../types"
import { mount, shallowMount } from "@vue/test-utils"
import { describe, expect, it } from "vitest"
import IngredientsList from "@/components/IngredientsList.vue"
import { createStoreMock } from "@/mocks/store"

const newIngredient = { id: "", amount: 3, unit: "Stk", name: "Aubergine" }
const existingIngredient = { id: "123", amount: 8, unit: "L", name: "abc" }

describe("IngredientsList", () => {
  it("should render an empty list without edit mode", () => {
    const wrapper = shallowMount(IngredientsList, getOptions([]))
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("should render a list with an item", () => {
    const wrapper = shallowMount(IngredientsList, getOptions([existingIngredient]))
    expect(wrapper.html()).toMatchSnapshot()
  })

  it.only("should render a list with an editable item and with a new item", () => {
    const wrapper = shallowMount(IngredientsList, getOptions([existingIngredient], true))
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("should add new items", async () => {
    const wrapper = shallowMount(IngredientsList, getOptions([existingIngredient], true))
    await wrapper.vm.$refs.newItem.$emit("input", newIngredient)
    expect(wrapper.emitted().input).toEqual([[[existingIngredient, newIngredient]]])
  })

  it("should change existing items", async () => {
    const wrapper = mount(IngredientsList, getOptions([existingIngredient], true))
    await wrapper.vm.$refs.existingItem[0].$emit("input", { ...existingIngredient, amount: 5 })
    expect(wrapper.emitted().input).toEqual([[[{ ...existingIngredient, amount: 5 }]]])
  })

  it("should remove existing items", async () => {
    const wrapper = mount(IngredientsList, getOptions([existingIngredient], true))
    await wrapper.vm.$refs.existingItem[0].$emit("delete")
    expect(wrapper.emitted().input).toEqual([[[]]])
  })
})

function getOptions(ingredients: Ingredient[], editable = false) {
  const { store } = createStoreMock({ ingredients })
  return { global: { plugins: [store] }, props: { ingredients, editable } }
}
