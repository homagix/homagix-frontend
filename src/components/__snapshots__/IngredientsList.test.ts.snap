// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`IngredientsList > should render a list with an editable item and with a new item 1`] = `"<div class=\\"ingredient-list\\" data-v-973b5ae8=\\"\\"><span data-v-973b5ae8=\\"\\">8</span><span data-v-973b5ae8=\\"\\">L</span><span data-v-973b5ae8=\\"\\">abc</span></div>"`;

exports[`IngredientsList > should render a list with an item 1`] = `"<div class=\\"ingredient-list\\" data-v-973b5ae8=\\"\\"><span data-v-973b5ae8=\\"\\">8</span><span data-v-973b5ae8=\\"\\">L</span><span data-v-973b5ae8=\\"\\">abc</span></div>"`;

exports[`IngredientsList > should render an empty list 1`] = `"<div class=\\"ingredient-list\\" data-v-973b5ae8=\\"\\"></div>"`;
